presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java

```diff
@@ -133,25 +134,76 @@ public void testOutOfMemoryKiller()
+                query.get();
+            }
+        }
+    }
+
+    private void waitForQueryToBeKilled(DistributedQueryRunner queryRunner)
+            throws InterruptedException
+    {
+        boolean queryDone = false;
+        while (!queryDone) {
+            for (BasicQueryInfo info : queryRunner.getCoordinator().getQueryManager().getQueries()) {
+                if (info.getState().isDone()) {
+                    assertNotNull(info.getErrorCode());
+                    assertEquals(info.getErrorCode().getCode(), CLUSTER_OUT_OF_MEMORY.toErrorCode().getCode());
+                    queryDone = true;
+                    break;
                 }
-                MILLISECONDS.sleep(10);
             }
+            MILLISECONDS.sleep(10);
+        }
+    }
 
-            // Release the memory in the reserved pool
+    @Test(timeOut = 240_000, expectedExceptions = ExecutionException.class, expectedExceptionsMessageRegExp = ".*Query killed because the cluster is out of memory. Please try again in a few minutes.")
+    public void testReservedPoolDisabled()
+            throws Exception
+    {
+        Map<String, String> properties = ImmutableMap.<String, String>builder()
+                .put("experimental.legacy-reserved-pool-enabled", "false")
+                .put("query.low-memory-killer.delay", "5s")
+                .put("query.low-memory-killer.policy", "total-reservation")
+                .build();
+
+        try (DistributedQueryRunner queryRunner = createQueryRunner(TINY_SESSION, properties)) {
+            // Reserve all the memory
+            QueryId fakeQueryId = new QueryId("fake");
+            for (TestingPrestoServer server : queryRunner.getServers()) {
+                List<MemoryPool> memoryPools = server.getLocalMemoryManager().getPools();
+                assertEquals(memoryPools.size(), 1, "Only general pool should exist");
+                assertTrue(memoryPools.get(0).tryReserve(fakeQueryId, "test", memoryPools.get(0).getMaxBytes()));
+            }
+
+            List<Future<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < 2; i++) {
+                queryFutures.add(executor.submit(() -> queryRunner.execute("SELECT COUNT(*), clerk FROM orders GROUP BY clerk")));
+            }
+
+            // Wait for one of the queries to die
+            waitForQueryToBeKilled(queryRunner);
+
+            // Reserved pool shouldn't exist on the workers and allocation should have been done in the general pool
             for (TestingPrestoServer server : queryRunner.getServers()) {
-                MemoryPool reserved = server.getLocalMemoryManager().getPool(RESERVED_POOL);
+                Optional<MemoryPool> reserved = server.getLocalMemoryManager().getReservedPool();
+                MemoryPool general = server.getLocalMemoryManager().getGeneralPool();
+                assertFalse(reserved.isPresent());
+                assertTrue(general.getReservedBytes() > 0);
```

Comment:
Is there a race here where one query is killed an the other finishes immediately, so the general pool reservation is empty?

Reply:
It's possible in theory, but unlikely as this query isn't likely to finish fast. Another test (`testOutOfMemoryKiller`) has a similar logic and we haven't observed a race in that test so far AFAIK.